Add AntennaPairFile for compressing antenna columns

aroffringa opened the pull request.

mickveldhuis reviewed (commented):
Looks good to me overall, just left a few small comments.

aroffringa commented on tables/AlternateMans/AntennaPairFile.h:
```diff
@@ -0,0 +1,245 @@
+    ReadData();
+  }
+
+  static bool HasUnsetAntenna(const std::array<int32_t, 2>& pair) {
```
I also always question what type to use ;) but I chose std::array because the type of the elements is enforced to be the same. A std::pair can have two different types (so the type would be std::pair<int32_t, int32_t>). Moreover, it doesn't allow iteration etc. over its elements because of that, and that's used in the templated code.

aroffringa commented on tables/AlternateMans/test/tAntennaPairFile.cc:
```diff
@@ -0,0 +1,186 @@
+  BOOST_CHECK_EQUAL(file.ReadAntenna1(0), 3);
+  BOOST_CHECK_THROW(file.ReadAntenna2(0), std::runtime_error);
+
+  // Rewriting a value with the same value should be allowed
```
I think this is a use-case, sometimes. There are tools that when rewriting the data also rewrite other columns; this is fine if the values are the same (which is likely). I don't think Dp3 does this when updating a ms though, so for our scenarios it might not be so relevant that it supports that.

aroffringa commented on tables/CMakeLists.txt:
```diff
@@ -339,6 +339,7 @@ DESTINATION include/casacore/tables/Tables
 )
 
 install (FILES
+AlternateMans/AntennaPairFile.h
```
The Stokes I data manager isn't entirely comparable because it implements the storage manager. The class added in this MR is "just" the file class that does the compression; a second MR will still have to add the add the storage manager class, which will be header+unit file like the Stokes I mgr.

If possible and appropriate, a header file is imho somewhat preferred because it saves code and puts everything in the same place. If this makes the header file very large, or if it pulls in many dependencies (that may even be circular), it might be preferred to split it in unit/header file. Inlined and templated code on the other hand often is required to be in the header (which in this case at least affects on of the biggests functions).

As far as I can find, google style doesn't really say anything about when to make header files or when not, so it's somewhat of a subjective balance.

mickveldhuis replied:
Ah okay, got it! That's fair.

aroffringa commented:
Maik made in the UvwFile MR the good comment to check if the header tag has the correct tag when opening a file. I'll apply that in this MR as well.

aroffringa commented:
> Maik made in the UvwFile MR the good comment to check if the header tag has the correct tag when opening a file. I'll apply that in this MR as well.

This is now added. This highlighted that if the file remains empty, the header was never written. This is now also fixed.

mickveldhuis commented on tables/AlternateMans/AntennaPairFile.h:
```diff
@@ -199,11 +199,17 @@ class AntennaPairFile {
     file_.ReadHeader(data);
+    if (!std::equal(data, data + 8, kMagicHeaderTag)) {
+      throw std::runtime_error(
+          "The Antenna-pair columnar file header not have the expected tag for "
```
I have a feeling there's a "does" missing in this sentence, shouldn't it be something like: "... file header does not have ..."?

aroffringa replied:
Yes indeed, good catch :)

aroffringa replied:
(now fixed)
